assets/js/base/utils/tracks.js

```diff
@@ -0,0 +1,25 @@
+/**
+ * External dependencies
+ */
+import { select } from '@wordpress/data';
+
+function isTracksAvailable() {
+	return typeof window.wcTracks.recordEvent === 'function';
+}
+
+export const recordEditorEvent = function( event, props ) {
+	if ( ! isTracksAvailable ) {
+		return;
+	}
+
+	// Force prefix - our editor events will be 'wcadmin_blocks_*'.
+	const blocksPrefix = 'blocks_';
+
+	const postData = select( 'core/editor' );
+	window.wcTracks.recordEvent( blocksPrefix + event, {
+		// Automatically include post id and post type props.
+		post_id: postData.getCurrentPostId(),
+		post_type: postData.getCurrentPostType(),
+		...props,
+	} );
+};
```

Comment:
Some things to consider here:

- This utility is tightly coupled to recording events in the context of the `core/editor` data store and the post/post-type values.  As we start building interfaces outside of the WP editor are we going to build utility functions like this for every interface? What if we want to track other things about the editor instance beyond post_id and post_type?
- This is using a data selector that is not subscribed. In most cases (like this one), this will probably be okay (because we only _want_ the values at the time of select). But we do need to keep in mind it will only select the data at the time of tracking the event.

With those considerations, I think it might be good to implement tracks as a custom React hook. You could have something like this:

```jsx
export const useTracks = ( event, trackCallback ) => {
  if ( typeof trackCallback !== 'function' ) {
    throw new Error( 'trackCallback must be a function' );
  }
  const recordEvent = useCallback( (...data ) => {
    const recordedData = trackCallback( ...data );
    if ( ! isTracksAvailable() ) {
      return;
    }
    window.wcTracks.recordEvent( `blocks_${ event }`, recordedData );
  }, [ event, trackCallback ] );
  return recordEvent;
}

// Example implementation of hook
const MyComponentWithTracks = ( props ) => {
  const recordEvent = useTracks("cart_shipping_calculator_toggled", ( trackedProps ) => {
    const postData = select( 'core/editor' );
    return {
      post_id: postData.getCurrentPostId(),
      post_type: postData.getCurrentPostType(),
      ...trackedProps
    };
  });
  /** where tracking happens later in component */
  recordEvent( { enabled: ! isShippingCalculatorEnabled } );
```

The `trackCallback` then can be something that is extracted out for re-usability if needed but the hook allows for composition of various track events.

Unless there are reasonable objections, I'd like us to try this route (custom hook) and see how it works. If it works well, we should create a tracks package in the wc-admin repo that exposes the hook (and possibly other tracks utils) for use by all Woo projects so that there's a common mechanism for implementing tracks in our projects.

Reply:
I had a quick look at setting up a custom hook, but it didn't seem to be offering anything useful. I can imagine this might be a useful approach as we iterate on this, but am struggling to justify this approach right now.

Responding to specific points:

> This utility is tightly coupled to recording events in the context of the core/editor data store and the post/post-type values. As we start building interfaces outside of the WP editor are we going to build utility functions like this for every interface? 

This is intentional - this API is just for Woo Blocks plugin for now, and for triggering events in the block editor (which is the bulk of what we need to do). I.e. scoped to this plugin and its needs for now.

I like the idea of generalising this to other use cases and exporting an API from wc-admin, but I'm not convinced we need to solve this / prototype this now. I'm focused on the immediate need of sending events for woo blocks editor features.

And just to clarify: the common props (e.g. `post_id`) and custom per-event props (`enabled`) in this PR are not set in stone, these are some examples to get us started. As we flesh out the events it will become clearer what consistent data we want on all block-editor events, vs what's custom per event. We may also want to add common props to all tracks events, further up the chain in Woo core (e.g. `url`, `woo_version`).

> As we start building interfaces outside of the WP editor are we going to build utility functions like this for every interface?

Yes ish – wc-admin already has a working solution for much of the rest of admin screens. This adds a specific wrapper to that for blocks editor events.

> This is using a data selector that is not subscribed. In most cases (like this one), this will probably be okay (because we only want the values at the time of select). But we do need to keep in mind it will only select the data at the time of tracking the event.

I don't see this as a problem – the event props should be populated with data that matches what's happening at the time of the event. Do you have an event or use-case in mind? With `cart_shipping_calculator_toggled` example, the props could be passed in at event time, or supplied as defaults, I don't think there's a need to supply these via callback.

I can imagine that patterns might emerge as we implement more events – perhaps we can iterate to a custom hook or other abstraction as we go.

Let's have a chat about this – I want to understand better what the advantages would be of using a hook (I may be missing something).

Reply:
Just looking at the function vs the hook suggestion, I immediately see a pro to the hook; `recordEditorEvent` forces the use of `const postData = select( 'core/editor' );`. If we ever want different data for the event you either have to edit this and affect all tracked events, or create more similar methods grabbing different sets of data.

With the hook, when you use it you can immediately override how the tracking event should work:

```
( trackedProps ) => {
    const postData = select( 'core/editor' );
    return {
      post_id: postData.getCurrentPostId(),
      post_type: postData.getCurrentPostType(),
      ...trackedProps
    };
  }
```
That's more flexible. Also to note, we wouldn't need to do this everywhere. We could have a default `trackCallback` that gets used when a custom one isn't provided.

Additionally, the suggested hook is currently just returning a single function:

```
return recordEvent;
```

This could be expanded to return other tracks related functions from the hook if we wanted. For example, one could create a wrapper for setAttributes and return that here to simplify some of the integration with components.

> As we flesh out the events it will become clearer what consistent data we want on all block-editor events, vs what's custom per event. We may also want to add common props to all tracks events, further up the chain in Woo core (e.g. url, woo_version).

That could be in the hook, given the suggested approach.

Reply:
Also to be clear, we could have a set of common track actions exported from the tracks package, for instance:

```js
const includePostdata = ( trackedProps ) => {
    const postData = select( 'core/editor' );
    return {
      post_id: postData.getCurrentPostId(),
      post_type: postData.getCurrentPostType(),
      ...trackedProps
    };
  };
```

And then implementation would be something like:

```jsx
import { useTracks, includePostData } from '@woocommerce/tracks';

const MyTrackedComponent = ( props ) => {
  const recordEvent = useTracks( 'cart_shipping_calculator_toggled', includePostData )
  const onClick = () => {
    recordEvent( { enabled: ! isShippingCalculatorEnabled } );
    props.onClick;
  }
  return <button onClick={ onClick }></button>;
}
```

This could even allow for composing various data we want included in the track in future iterations:

```jsx
const recordEvent = useTracks( 'some_event_slug', [ 'includePostData', 'includeSiteData' ] );
```

Reply:
Here's an exploration of the hook approach. https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/try/woo-blocks-editor-tracks-events-hook

It looks to me like we'd be better off just providing useful `select` utils like `selectPostInfoProps()`. 

One specific thing I want to get in place now is a way to standardise common props for all woo blocks editor events - wrapping `wcTracks.recordEvent` seems the simplest way to do that.

Reply:
Thanks for doing the exploration of the hook approach.  

> If we go that way, it looks to me like we'd be better off just providing useful select utils like selectPostInfoProps()

That is basically what `includePostData` was in my examples above.

> the hook isn't providing anything particularly useful.

Can you expand on this please? I thought Mike and I gave some reasons why the hook has value (both for the work in this pull and also for needed future use as we expand out the implementation of data tracking) but I don't see where you've addressed that feedback.